device/iot-device-samples/device-connection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/DeviceClientManager.java

```diff
@@ -0,0 +1,415 @@
+        {
+            for (String propertyKey : this.reportedPropertiesToSend.keySet())
+            {
+                System.out.println("Failed to update reported properties with new key " + propertyKey + " with value " + this.reportedPropertiesToSend.get(propertyKey) + " due to retryable error with status code " + statusCode.name() + ". Will try sending again later.");
+            }
+        }
+        else
+        {
+            String newReportedPropertyValue = (String) twin.getReportedProperties().remove(newReportedPropertyKey);
+            System.out.println("Failed to update reported properties with new key " + newReportedPropertyKey + " with value " + newReportedPropertyValue + " due to an unretryable error with status code " + statusCode.name() + ". Removing new property from twin.");
+        }
+    }
+
+    // callback for when a desired property update is received
+    @Override
+    public void onDesiredPropertiesUpdated(Twin newTwin, Object context)
+    {
+        System.out.println("Desired properties update received by device");
+        this.twin.getDesiredProperties().putAll(newTwin.getDesiredProperties());
+        this.twin.getDesiredProperties().setVersion(newTwin.getDesiredProperties().getVersion());
+
+        System.out.println("New twin state from device side:");
+        System.out.println(this.twin.toString());
+    }
+
+    // callback for when a get twin call has succeeded and the current twin has been retrieved.
+    @Override
+    public void onTwinReceived(Twin twin, IotHubClientException e, Object context)
+    {
+        if (e == null)
+        {
+            System.out.println("Received the current twin state: ");
+            System.out.println(twin.toString());
+            this.twin = twin;
+            this.reportedPropertiesToSend.setVersion(twin.getReportedProperties().getVersion());
+            this.gettingTwinAfterReconnection = false;
+        }
+        else if (e.isRetryable())
+        {
+            System.out.println("Encountered a retryable error with status code while trying to get the client's twin. Trying again...");
+            this.deviceClient.getTwinAsync(this, null);
+        }
+        else
+        {
+            System.out.println("Encountered a non retryable error while trying to get the client's twin. Abandoning getting twin.");
+        }
+
+    }
+    // endregion
+
+    // region direct methods
+    // callback for when a direct method is invoked on this device
+    @Override
+    public DirectMethodResponse onMethodInvoked(String methodName, DirectMethodPayload payload, Object context)
+    {
+        System.out.println("Method " + methodName + " invoked on device.");
+        return new DirectMethodResponse(200, null);
+    }
+    // endregion
+
+    // region cloud to device messaging
+    // callback for when a cloud to device message is received by this device
+    @Override
+    public IotHubMessageResult execute(Message message, Object callbackContext)
```

Comment:
Can we find a better method name for this, perhaps something like `onC2dMessageReceived`?